### PR TITLE
Implement assert_implements and assert_implements_optional

### PR DESCRIPTION
**Pull request by stephenmcgruer**
See https://github.com/web-platform-tests/rfcs/pull/48 and
https://github.com/web-platform-tests/wpt/issues/21971.

Two tests are ported to the new methods to show their use.

Note that this PR does not remove the deprecated `assert_precondition` -
that will be done in follow-up PRs.

### CODE REVIEW

**Comment by stephenmcgruer** [comment]
I couldn't figure out if there's a place for unittests for testharness.js that aren't expected to fail. Do we have some somewhere?

**Comment by stephenmcgruer** [comment]
(TaskCluster didn't pick up the moving from draft --> review, so close/open to kick it)

File: infrastructure/expected-fail/assert_implements_optional.html [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<!DOCTYPE html>
```
Comment (jgraham) [inline infrastructure/expected-fail/assert_implements_optional.html:1]: I think the logical place to put this kind of test is under `resources/tests` (although those are sadly more work to write an run, but they are also enabled in CI so just as effective for preventing regressions)

Comment (stephenmcgruer) [reply infrastructure/expected-fail/assert_implements_optional.html:1]: > I think the logical place to put this kind of test is under resources/tests (although those are sadly more work to write an run, but they are also enabled in CI so just as effective for preventing regressions)

That looks like exactly what I want, awesome!

qq; it looks like `functional/` exists to test harness failure cases, but there are also `unit/` tests like `assert_object_equals.html` which use the fact that `assert()` throws an `AssertionError` to create failure tests without needing a comparison output:

```
function test_failure(fn, name) {
  test(function() {
    try {
      fn();
    } catch (err) {
      if (err instanceof AssertionError) {
        return;
      }
      throw new AssertionError('Expected an AssertionError, but');
    }
    throw new AssertionError(
      'Expected an AssertionError, but no error was thrown'
    );
  }, name);
}
```

What do you think of this pattern? I quite like it, but don't want to put the effort into using it if you're then going to ask me to switch to `functional/` for the failing tests :). (And I'm happy to use `functional/` if that's what we prefer).

Comment (jgraham) [reply infrastructure/expected-fail/assert_implements_optional.html:1]: I don't have a strong preference. I think the format in `functional` turned out to be over-verbose, so I can understand not wanting to use that, but it is arguably more consistent. Up to you.

Comment (stephenmcgruer) [reply infrastructure/expected-fail/assert_implements_optional.html]: Done. PTAL :)

**Comment by stephenmcgruer** [comment]
Ping @jgraham :)

**Review by jgraham — APPROVED** [review]

**Comment by stephenmcgruer** [comment]
(Sigh, and I forget again that GitHub isn't Gerrit, and doesn't update the commit description with the PR description. It's only slightly inaccurate - commit claims it doesn't port any tests but it actually changes two...)

**Pull request merged**